crates/matrix-sdk-crypto/src/store/mod.rs

```diff
@@ -204,15 +200,29 @@ impl<'a> SyncedKeyQueryManager<'a> {
         user: &UserId,
     ) -> UserKeyQueryResult {
         let mut users_for_key_query = self.manager.users_for_key_query.lock().await;
-
         let Some(waiter) = users_for_key_query.maybe_register_waiting_task(user) else {
             return UserKeyQueryResult::WasNotPending;
         };
 
         let wait_for_completion = async {
-            while !waiter.completed.load(Ordering::Relaxed) {
-                users_for_key_query =
-                    self.manager.users_for_key_query_condvar.wait(users_for_key_query).await;
+            loop {
+                // Register for being notified before releasing the mutex, so
+                // it's impossible to miss a wakeup between the last check for
+                // whether we should wait, and starting to wait.
+                let mut notified = pin!(self.manager.users_for_key_query_notify.notified());
+                notified.as_mut().enable();
+                drop(users_for_key_query);
+
+                // Wait for a notification
+                notified.await;
+
+                // Reclaim the lock before checking the flag to avoid races
+                // when two notifications happen right after each other and the
+                // second one sets the flag we want to wait for.
+                users_for_key_query = self.manager.users_for_key_query.lock().await;
+                if waiter.completed.load(Ordering::Relaxed) {
+                    break;
+                }
```

Comment:
I think having this at the top of the loop was not only clearer, but potentially faster (it will skip an unlock/wait/lock cycle in the (maybe unlikely) case that the update arrives during/just after the call to `maybe_register_waiting_task`).  But it's not important.

Reply:
Ah yes, now this is just a do-while loop. I wouldn't do it for performance, but I'll move back to a `while` loop so we have a smaller diff / the code is a bit easier to read.